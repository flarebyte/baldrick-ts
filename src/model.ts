export interface PackageJson {
  name: string;
  description: string;
  keywords: string[];
  author: Author | string;
  version: string;
  license: string;
  homepage: string;
  repository: Repository;
  main: string;
  typings: string;
  files: string[];
  engines: Engines;
  scripts: Scripts;
  module: string;
  devDependencies: Dependencies;
  dependencies: Dependencies;
  peerDependencies: Dependencies;
}

export interface PackageKeyStats {
  key: string;
  countItems: number;
  stringLength: number;
}

export interface Author {
  name: string;
  url: string;
}

export interface Dependencies {
  [key: string]: string;
}

export interface Engines {
  node: string;
}

export interface Repository {
  type: string;
  url: string;
}

export interface Scripts {
  [key: string]: string;
}

export enum LicenseType {
  MIT,
  Other,
}

export enum ScaffoldingType {
  TsDx,
  Other,
}

export enum PipelineType {
  Github,
  Other,
}

export enum ProjectType {
  TsLib,
  TsCli,
  Other,
}

export interface CoreProject {
  name: string;
  githubAccount: string;
  licenseType: LicenseType;
  scaffoldingType: ScaffoldingType;
  pipelineType: PipelineType;
  projectType: ProjectType;
}

export interface TodoPriority {
  name: string;
  specialChar: string;
}
export interface Todo {
  name: string;
  description: string;
  priority: TodoPriority;
}

export enum FieldStatus {
  Ok,
  Todo,
  Fixable,
}
export type toFieldStatus = (value: any) => FieldStatus;

export interface PackageJsonStatusConverter {
  name: toFieldStatus;
  description: toFieldStatus;
  keywords: toFieldStatus;
  author: toFieldStatus;
  version: toFieldStatus;
  license: toFieldStatus;
  homepage: toFieldStatus;
  repository: toFieldStatus;
  main: toFieldStatus;
  typings: toFieldStatus;
  files: toFieldStatus;
  engines: toFieldStatus;
  scripts: toFieldStatus;
  module: toFieldStatus;
  devDependencies: toFieldStatus;
  dependencies: toFieldStatus;
  peerDependencies: toFieldStatus;
}

export interface PackageJsonStatus {
  name: FieldStatus;
  description: FieldStatus;
  keywords: FieldStatus;
  author: FieldStatus;
  version: FieldStatus;
  license: FieldStatus;
  homepage: FieldStatus;
  repository: FieldStatus;
  main: FieldStatus;
  typings: FieldStatus;
  files: FieldStatus;
  engines: FieldStatus;
  scripts: FieldStatus;
  module: FieldStatus;
  devDependencies: FieldStatus;
  dependencies: FieldStatus;
  peerDependencies: FieldStatus;
}
