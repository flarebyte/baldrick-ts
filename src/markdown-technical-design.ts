import { markdownToString, parseMarkdown } from './markdown.js';
import { CoreProject, MdSection } from './model.js';

const keepSections = (section: MdSection): boolean =>
  ['Overview', 'Purpose'].includes(section.title);

const codeStructure: MdSection = {
  title: 'Code structure',
  body: [
    '* __src__: Typescript source code',
    '* __test__: Jest unit tests',
    '* __dist__: Temporary folder for building distribution code',
    '* __report__: Temporary folder for reporting; usually for continuous integration',
    '* __.github__: Folder for github pipeline',
    '* __.vscode: Folder for visual code snippets',
  ].join('\n'),
};

const docAndLinks: MdSection = {
  title: 'Useful links',
  body: [
    '* Guideline for [Clean Code in Typescript](https://labs42io.github.io/clean-code-typescript/)',
  ].join('\n'),
};

export const toTechnicalDesignMd = (
  _core: CoreProject,
  existingMd: string
): string => {
  const existing = parseMarkdown(existingMd);

  const sections = [
    ...existing.sections.filter(keepSections),
    codeStructure,
    docAndLinks,
  ];
  const updated = {
    ...existing,
    title: 'Technical Design',
    sections,
  };
  const rawReadme = markdownToString(updated);
  return rawReadme.replace(/[\r\n]{3,}/gm, '\n\n');
};
