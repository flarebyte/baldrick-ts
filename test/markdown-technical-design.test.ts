import { libCoreProject } from './fixture-core-project.js';
import { CoreProject } from '../src/model.js';
import { toTechnicalDesignMd } from '../src/markdown-technical-design.js';

const exampleReadme = `
# Technical Design

## Overview

### Audience

### Purpose

### Code structure

### Useful links

`;

describe('Technical Design documentation', () => {
  it('Updates TECHNICAL_DESIGN.md with standardized chapters', () => {
    const project: CoreProject = {
      ...libCoreProject,
      name: 'project123',
    };
    const actual = toTechnicalDesignMd(project, exampleReadme);
    expect(actual).toMatchSnapshot();
  });
});
