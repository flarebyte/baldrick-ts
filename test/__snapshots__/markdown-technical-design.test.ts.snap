// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Technical Design documentation Updates TECHNICAL_DESIGN.md with standardized chapters 1`] = `
"# Technical Design

> 

## Overview

### Audience

### Purpose

### Code structure

### Useful links

## Code structure

* __src__: Typescript source code
* __test__: Jest unit tests
* __dist__: Temporary folder for building distribution code
* __report__: Temporary folder for reporting; usually for continuous integration
* __.github__: Folder for github pipeline
* __.vscode__: Folder for visual code snippets

## Useful links

* Guideline for [Clean Code in Typescript](https://labs42io.github.io/clean-code-typescript/)"
`;
